src/main/kotlin/org/rust/ide/annotator/RsErrorAnnotator.kt

```diff
@@ -884,6 +885,33 @@ class RsErrorAnnotator : AnnotatorBase(), HighlightRangeExtension {
         RsDiagnostic.LoopOnlyKeywordUsedOutsideOfLoopError(expr.operator).addToHolder(holder)
     }
 
+    private fun checkArrayType(holder: RsAnnotationHolder, o: RsArrayType) {
+        collectDiagnostics(holder, o)
+        val sizeExpr = o.expr
+        if (sizeExpr != null && o.arraySize == null) {
+            checkArraySizeExpr(holder, sizeExpr)
+        }
+    }
+    private fun checkArrayExpr(holder: RsAnnotationHolder, o: RsArrayExpr) {
+        val sizeExpr = o.sizeExpr
+        if (sizeExpr != null) {
+            checkArraySizeExpr(holder, sizeExpr)
+        }
+    }
+
+    private fun checkArraySizeExpr(holder: RsAnnotationHolder, sizeExpr: RsExpr) {
+        sizeExpr.descendantsOfTypeOrSelf<RsPathExpr>().forEach { pathExpr ->
+            val ref = pathExpr.path.reference?.resolve()
+            if (ref != null && ref !is RsConstant && (ref as? RsFunction)?.isConst != true) {
+                if (pathExpr.parent is RsCallExpr) {
+                    RsDiagnostic.NonConstantFunctionCallInConstantError(pathExpr).addToHolder(holder)
+                } else {
+                    RsDiagnostic.NonConstantValueInConstantError(pathExpr).addToHolder(holder)
+                }
```

Comment:
Looks like it's slightly better because allow not to distinguish function calls (`RsCallExpr`) from method calls (`RsMethodCall`)
```suggestion
                val diagnostic = if (ref is RsPatBinding) {
                    RsDiagnostic.NonConstantValueInConstantError(pathExpr)    
                } else {
                    RsDiagnostic.NonConstantFunctionCallInConstantError(pathExpr)
                }
                diagnostic.addToHolder(holder)
```